### PR TITLE
Plans: Update "free domain with plan" sidebar upsell to go to domains first

### LINKED ISSUE #66017: Domain Upsell Nudge: redirect users to domain search first
In #51609, we updated the "Free domain with annual plan" sidebar upsell nudge to direct users to pick a plan before searching for a domain (instead of adding one silently in the background). This was to address user confusion and refund requests.

Since the goal of the upsell nudge is to get a domain, let's update the flow to:

1. Click nudge
2. See Domains search page
3. If a domain is added, see Plans page (skip email upsell for now)
4. If plan is selected, go to Checkout

We might need to see this flow in action to see if any refinements are needed.

### PR DESCRIPTION
**Pull request by sirbrillig**
#### Proposed Changes

In https://github.com/Automattic/wp-calypso/pull/51609, we updated the "Free domain with annual plan" sidebar upsell nudge to direct users to pick a plan before searching for a domain (instead of adding one silently in the background). This was to address user confusion and refund requests. However, since the primary goal of users clicking on that upsell is to get a domain, the intervening plans page can itself be confusing.

**The current flow is: Upsell button > Plans page > Domain search > (maybe) Email upsell > Checkout.**

In this PR we modify the upsell flow so that we go to the domains search page first, then the plans page, then directly to checkout. 

**The new flow is: Upsell button > Domain search > Plans page > Checkout.**

There are a few other differences as well:

1. An informational banner will be displayed on the plans page to explain that we are in the upsell flow (see the screenshot below).
2. The email upsell after selecting a domain is skipped in this flow to simplify the steps as requested in https://github.com/Automattic/wp-calypso/issues/66017. 

Some possible problems with this change:

- The plans page supports a `redirectTo` prop that would be used by checkout after the product has been added. This flow will ignore that prop, but since it is added by the `?redirect_to=` query string and this flow is started from a hard-coded link I don't think they will ever be combined.
- The plans page supports a `withDiscount` prop to add a coupon to the cart along with the plan. This flow will ignore that prop, but since it is added by the `?discount=` query string and this flow is started from a hard-coded link I don't think they will ever be combined.
- If you select a domain transfer or domain mapping from the domains search page, you'll effectively leave the upsell flow and will not end up on the plans page when adding the domain product is complete.

Fixes https://github.com/Automattic/wp-calypso/issues/66017

This must be merged first, but the change will not be active until D84928-code is merged to change where the upsell link goes.

#### Screenshots

Screen Shot 2022-07-27 at 12 13 11 PM

Screen Shot 2022-07-28 at 6 03 36 PM

Screen Shot 2022-07-27 at 1 37 27 PM

#### Testing Instructions

- First, apply D84928-code and sandbox the API.
- Start with a site that has no paid plan.
- Visit the customer home page for the site, eg: http://calypso.localhost:3000/home/example.com
- Click the "Upgrade" button in the sidebar notification that reads "Free domain with an annual plan" (see first screenshot above).
- Verify that you are redirected to the domains page for your site (see second screenshot above).
- Click to select a domain.
- Verify that you are redirected to the plans page (see third screenshot above).
- Click the "Upgrade" button on a plan's card to select that plan.
- Verify that you arrive at checkout with both the plan and the domain in your cart.

### CODE REVIEW

**Comment by michaeldcain** [comment]
Thanks for the quick turnaround on this - the flow feels better.

> If a plan is not selected, the domain will not be added to the cart either; they are added together at the end of the flow. This avoids the original issue where users would end up aborting the flow and have an unexpected product in their cart.

I think this is slightly different than the original issue, where we were adding a plan to the cart in the background, and the user never explicitly knew that they had added it. I think it's completely reasonable to add the domain to the cart when they select the domain - especially if the domain CTA says "add to cart" or something like that. I guess I could go either way - we could wait for feedback from the P2 post.

**Comment by sirbrillig** [comment]
Per @michaeldcain's suggestions, I made the following changes (now reflected in the description and screenshots):

1. Clicking to add the domain will now add it to the cart immediately rather than waiting until the end of the flow.
2. The domain search page will now display "Free for the first year" next to most domain options.

**Comment by dzver** [comment]
This feels like a major improvement! Given that the plan is not auto-added, the conversion rate of the link might not be as good as it previously was, but it should go up. We can monitor the stats after the change is merged.

It looks good to me but I'll leave the final review to be from a person from Payments. I checked the code, and also tested it by following the test plan and trying a couple of other things that are not on the test plan. 

I also think this is safe to ship without an experiment, as long as we do a follow-up data request or analysis and see what happened.

**Review by michaeldcain — APPROVED** [review]
This works well for me, and is a definite improvement.

I tested the main flow (CTA > Domain > Plan > Checkout) and everything worked correctly. I also tested exiting the flow after picking a domain (correctly added to cart), mapping and transferring domains (this forked to the existing flows smoothly), and exiting before selecting a domain (no items in cart).

I updated the plans banner to use a title and hopefully reuse existing translation strings. If you don't like the look of it, you can revert to how you had it:

Screen Shot 2022-07-29 at 1 26 16 PM

**Pull request merged**
